plugins/NebulaTextures/src/SkyCoords.cpp

```diff
@@ -0,0 +1,212 @@
+ *   - CD1_1, CD1_2, CD2_1, CD2_2: The linear transformation matrix elements that map pixel coordinates to celestial coordinates.
+ *
+ * Returns:
+ *   - A QPair containing the longitude and latitude corresponding to the input pixel coordinates.
+ *
+ * * * * * * * * * * * * * * * * * * * *
+ *
+ * This function is based on the algorithm and approach from the code licensed under the Apache License, Version 2.0.
+ * The original code can be found at:
+ * https://github.com/PlanetaryResources/NTL-Asteroid-Data-Hunter/blob/master/Algorithms/Algorithm%20%231/alg1_psyho.h
+ *
+ * Copyright (C) [Year] [Original Authors].
+ *
+ * The function implementation is rewritten based on the logic and approach of the original code,
+ * but the code itself has been modified and is not directly copied.
+ *
+ * The code in this function is used under the terms of the Apache License, Version 2.0.
+ * You may not use the original code except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * If you modified this code, include a description of the changes made below:
+ * [Optional: Describe any modifications made]
+ */
+QPair<double, double> SkyCoords::pixelToRaDec(int X, int Y,
+	double CRPIX1, double CRPIX2,
+	double CRVAL1, double CRVAL2,
+	double CD1_1, double CD1_2,
+	double CD2_1, double CD2_2)
+{
+	// Convert the reference values (RA, Dec) to radians
+	double RA0 = CRVAL1 * D2R;
+	double Dec0 = CRVAL2 * D2R;
+
+	// Calculate dx and dy
+	double dx = X - CRPIX1;
+	double dy = Y - CRPIX2;
+
+	// Calculate xx and yy
+	double xx = CD1_1 * dx + CD1_2 * dy;
+	double yy = CD2_1 * dx + CD2_2 * dy;
+
+	// Calculate the celestial coordinates
+	double px = std::atan2(xx, -yy) * R2D;
+	double py = std::atan2(R2D, std::sqrt(xx * xx + yy * yy)) * R2D;
+
+	// Calculate sin(Dec) and cos(Dec)
+	double sin_dec = std::sin(D2R * (90.0 - CRVAL2));
+	double cos_dec = std::cos(D2R * (90.0 - CRVAL2));
+
+	// Calculate longitude offset (dphi)
+	double dphi = px - 180.0;
+
+	// Calculate celestial longitude and latitude
+	double sinthe = std::sin(py * D2R);
+	double costhe = std::cos(py * D2R);
+	double costhe3 = costhe * cos_dec;
+	double costhe4 = costhe * sin_dec;
+	double sinthe3 = sinthe * cos_dec;
+	double sinthe4 = sinthe * sin_dec;
+	double sinphi = std::sin(dphi * D2R);
+	double cosphi = std::cos(dphi * D2R);
+
+	// Calculate celestial longitude
+	double x = sinthe4 - costhe3 * cosphi;
+	double y = -costhe * sinphi;
+	double dlng = R2D * std::atan2(y, x);
+	double lng = CRVAL1 + dlng;
+
+	// Normalize the celestial longitude
```

Comment:
I meant that this, line 109-117 can go away as it's redundant with the following mod operation. And with fmodpos used in line 119, you can delete lines 120-122 as well.